[Merged by Bors] - feat(data/*, order/*) supporting lemmas for characterising well-founded complete lattices

ocfnash opened the pull request:
---
Note that I have included some extra lemmas even though they are not actually needed (or used). Here is the list of such lemmas together with my justification for including them:

- `finset.not_nonempty_iff_eq_empty`: seems to be a gap `set.not_nonempty_iff_eq_empty` already exists
- `finset.eq_singleton_iff_nonempty_unique_mem`: I introduce `set.eq_singleton_iff_nonempty_unique_mem` here so it seems worth adding the `finset` version also
- `finset.inf_eq_Inf`: I introduce `finset.sup_eq_Sup` so it seems worth adding the `inf` version also
- `rel_hom.map_inf`: I introduce `rel_hom.map_sup` here so it seems worth adding the `inf` version also
- `Inf_eq_top_nonempty_eq_singleton`: I introduce `Sup_eq_bot_nonempty_eq_singleton` so it seems worth adding the `inf` version also
- `set.finite.preimage_embedding`: I used this for an earlier proof of the results here and it seemed worth keeping

I think it is reasonable to bundle these extra results with the other work here but I will happily excise these (or move them to a separate PR) if requested.

Commit:
feat(src/order/complete_well_founded): characterise well-foundedness for complete lattices

awainverse commented on src/order/complete_well_founded.lean:
```diff
@@ -0,0 +1,156 @@
+end
+
+/-- Surely this exists somewhere (at least implicitly) already. -/
+lemma nat.finite_subset_is_bounded {s : set ℕ} (h : s.finite) : ∃ n, ∀ x ∈ s, x ≤ n :=
```
I bet you can do this for any `semilattice_sup` with `finset.sup`.

ocfnash replied:
Great point, you are of course correct:
```lean
lemma finite_subset_is_bounded {β : Type*} [semilattice_sup_bot β] {s : set β} (h : s.finite) :
  ∃ y, ∀ x ∈ s, x ≤ y :=
begin
  use h.to_finset.sup id,
  intros x hx, exact finset.le_sup (set.finite.mem_to_finset.mpr hx),
end
```

Thanks for such helpful remarks here and elsewhere. I finally have time again this week and plan to rework this PR substantially tomorrow, using your suggestions.

awainverse commented on src/order/complete_well_founded.lean:
```diff
@@ -0,0 +1,156 @@
+
+/-- A compactness property for a complete lattice is that any subset has a finite subset with the
+same `Sup`. -/
+def is_Sup_finite_compact := ∀ (s : set α), ∃ (t ⊆ s) (h : t.finite), Sup s = Sup t
```
Should there be another version with `t : finset α` and `finset.sup` instead of `Sup`?

awainverse replied:
Actually, I'd probably only use the finset version.

Commit:
Generalise `finite_subset_is_bounded`

Commit:
Improve proofs

Still quite a bit of work to do but proofs reasonable now.

Commit:
Generalise `nat_rel_embedding_map_sup` and improve module doc

Commit:
Add lemma `Inf_eq_top_nonempty_eq_singleton`

Just for completeness since I have already added `Sup_eq_bot_nonempty_eq_singleton`

eric-wieser commented on src/data/finset/lattice.lean:
```diff
@@ -121,13 +122,27 @@ calc t.sup f = (s ∪ t).sup f : by rw [finset.union_eq_right_iff_subset.mpr hst
+    cases t.eq_empty_or_nonempty with hte htne,
+    { subst hte, simp only [insert_emptyc_eq, id.def, finset.sup_singleton, h₃], },
+    { rw [finset.sup_insert, id.def], exact h h₃.1 (h₁ htne h₃.2), }, },
+end
```
Possibly tidier with `induction`:
```suggestion
lemma sup_closed_of_sup_closed {s : set α} (t : finset α) (htne : t.nonempty) (h_subset : ↑t ⊆ s)
  (h : ∀⦃a b⦄, a ∈ s → b ∈ s → a ⊔ b ∈ s) : t.sup id ∈ s :=
begin
  classical,
  induction t using finset.induction_on with x t h₀ h₁,
  { exfalso, apply finset.not_nonempty_empty htne, },
  { rw [finset.coe_insert, set.insert_subset] at h_subset,
    cases t.eq_empty_or_nonempty with hte htne,
    { subst hte, simp only [insert_emptyc_eq, id.def, finset.sup_singleton, h_subset], },
    { rw [finset.sup_insert, id.def], exact h h_subset.1 (h₁ htne h_subset.2), }, },
end
```

ocfnash replied:
Much nicer 😍

Commit:
Update src/data/finset/lattice.lean

Co-authored-by: Eric Wieser <wieser.eric@gmail.com>

eric-wieser commented on src/order/complete_lattice.lean:
```diff
@@ -266,9 +266,17 @@ iff.intro
 
+lemma Inf_eq_top_nonempty_eq_singleton {s : set α}
+  (h_inf : Inf s = ⊤) (hne : s.nonempty) : s = {⊤} :=
+by { rw set.eq_singleton_iff_nonempty_unique_mem, rw Inf_eq_top at h_inf, exact ⟨hne, h_inf⟩, }
```
Golfs to
```suggestion
set.eq_singleton_iff_nonempty_unique_mem.mpr ⟨hne, Inf_eq_top.mp h_inf⟩
```
I'm sure you can do something similar with the one below

ocfnash replied:
I have mixed feelings about this one. I do use term-mode proofs but whenever I'm on the fence, I avoid them because they're harder to read (and, I claim, maintain).

Still, commit away if you like!

urkud commented on src/order/complete_lattice.lean:
```diff
@@ -266,9 +266,17 @@ iff.intro
 @[simp] theorem Sup_eq_bot : Sup s = ⊥ ↔ (∀a∈s, a = ⊥) :=
 @Inf_eq_top (order_dual α) _ _
 
+lemma Sup_eq_bot_nonempty_eq_singleton {s : set α}
```
`eq_singleton_bot_of_Sup_eq_bot_of_nonempty`?

urkud commented on src/order/complete_well_founded.lean:
```diff
@@ -0,0 +1,88 @@
+  { rw [← ht₂, finset.sup_eq_Sup], exact Sup_le_Sup ht₁, },
+end
+
+lemma is_sup_closed_compact_of_is_Sup_finite_compact (h : is_Sup_finite_compact α) :
```
```suggestion
lemma is_Sup_finite_compact.is_sup_closed_compact (h : is_Sup_finite_compact α) :
```

urkud commented on src/order/complete_well_founded.lean:
```diff
@@ -0,0 +1,88 @@
+  { rw ht₂, exact t.sup_closed_of_sup_closed h ht₁ hsc, },
+end
+
+lemma well_founded_of_is_sup_closed_compact (h : is_sup_closed_compact α) :
```
```suggestion
lemma is_sup_closed_compact.well_founded (h : is_sup_closed_compact α) :
```

urkud commented on src/order/rel_iso.lean:
```diff
@@ -79,6 +79,30 @@ end
 | f ⟨H⟩ := ⟨λ a, f.acc _ (H _)⟩
 
+lemma monotone [partial_order α] [preorder β]
+  (f : ((<) : α → α → Prop) →r ((<) : β → β → Prop)) : monotone f :=
```
This is `strict_mono.monotone`.

ocfnash replied:
Aha, thanks!

Commit:
Update src/order/complete_well_founded.lean

Co-authored-by: Yury G. Kudryashov <urkud@urkud.name>

Commit:
Update src/order/complete_well_founded.lean

Co-authored-by: Yury G. Kudryashov <urkud@urkud.name>

Commit:
Drop 'rel_iso.monotone` in view of extant `strict_mono.monotone`.

Commit:
Make naming consistent after CR suggestions

Commit:
Naming suggestion from CR

Commit:
Add tfae lemma

Commit:
Add convenience `iff` lemmas for tfae result

Commit:
Add aliases for remaining implications

Commit:
Revert addition of new file `complete_well_founded.lean`

As per @urkud 's request, I am splitting this PR into two pieces:
  * One with all changes to existing files (this branch)
  * One with the proposed new file (a new branch).
